test/ectest.c

```diff
@@ -1556,6 +1556,145 @@ static const unsigned char p521_explicit[] = {
+        || !TEST_ptr(gtest = EC_GROUP_dup(group))
+        || !TEST_ptr(group_p = BN_new())
+        || !TEST_ptr(group_a = BN_new())
+        || !TEST_ptr(group_b = BN_new())
+        || !TEST_ptr(group_cofactor = BN_new())
+        || !TEST_ptr(group_gen = EC_GROUP_get0_generator(group))
+        || !TEST_ptr(group_order = EC_GROUP_get0_order(group))
+        || !TEST_true(EC_GROUP_get_cofactor(group, group_cofactor, NULL))
+        || !TEST_true(EC_GROUP_get_curve(group, group_p, group_a, group_b, NULL))
+        || !TEST_ptr(other_gen = EC_POINT_dup(group_gen, group))
+        || !TEST_true(EC_POINT_add(group, other_gen, group_gen, group_gen, NULL))
+        || !TEST_ptr(other_order = BN_dup(group_order))
+        || !TEST_true(BN_add_word(other_order, 1))
+        || !TEST_ptr(other_a = BN_dup(group_a))
+        || !TEST_true(BN_add_word(other_a, 1))
+        || !TEST_ptr(other_b = BN_dup(group_b))
+        || !TEST_true(BN_add_word(other_b, 1))
+        || !TEST_ptr(other_cofactor = BN_dup(group_cofactor))
+        || !TEST_true(BN_add_word(other_cofactor, 1)))
+        goto err;
+    field_nid = EC_METHOD_get_field_type(EC_GROUP_method_of(group));
+    if (field_nid == NID_X9_62_characteristic_two_field) {
+        if (!TEST_ptr(other_p = BN_dup(group_p))
+            || !TEST_true(BN_lshift1(other_p, other_p)))
+            goto err;
+    } else {
+        if (BN_num_bytes(group_order) > 28) {
+            if (!TEST_ptr(other_p = BN_dup(BN_get0_nist_prime_224())))
+                goto err;
+        } else {
+            if (!TEST_ptr(other_p = BN_dup(BN_get0_nist_prime_256())))
+                goto err;
+        }
+    }
+
+    /* Passes because this is a valid curve */
+    if (!TEST_int_eq(EC_GROUP_check_named_curve(group, 0), nid)
+        /* Only NIST curves pass */
+	|| !TEST_int_eq(EC_GROUP_check_named_curve(group, 1),
+                        EC_curve_nid2nist(nid) != NULL ? nid : NID_undef))
+        goto err;
+    /*
+     * Pass if the named curve is not known but the parameters are correct.
+     * It is possible to find the wrong alias if there is no curve name.
+     */
+    EC_GROUP_set_curve_name(group, NID_undef);
+    if (!TEST_int_gt(EC_GROUP_check_named_curve(group, 0), 0))
+        goto err;
+    /* Fail if the curve name doesnt match the parameters */
+    EC_GROUP_set_curve_name(group, nid + 1);
+    if (!TEST_int_eq(EC_GROUP_check_named_curve(group, 0), 0))
+        goto err;
+    EC_GROUP_set_curve_name(group, NID_undef);
+
+    /*
+     * If the built in curve has a seed and we set the seed to another value
+     * then it will fail the check. If the built in curve doesnt have a seed
+     * then setting the seed will pass the check (as the seed is optional).
+     */
+    has_seed = (EC_GROUP_get_seed_len(group) > 0);
+    if (!TEST_int_eq(EC_GROUP_set_seed(group, p521_named, 1), 1)
+        || !TEST_int_ne(EC_GROUP_check_named_curve(group, 0), has_seed))
+        goto err;
```

Comment:
Uhm, it seems a bit arbitrary to depend on the value of the first byte of `p521_named` as an invalid seed.
Also `p521_named` is completely unrelated to this test and it makes it a bit unreadable IMO.

What do you think about doing something like this instead?

~~~C
static int check_named_curve(int id)
{
    static const char *invalid_seed = "THIS IS NOT A VALID SEED";
    static size_t invalid_seed_len = strlen(invalid_seed);
    // [...]
    // at line 1633
    has_seed = (EC_GROUP_get_seed_len(group) > 0);
    if (!TEST_int_eq(EC_GROUP_set_seed(group, invalid_seed, invalid_seed_len), invalid_seed_len))
        goto err;
    if (has_seed) {
        /* If the built in curve has a seed and we set the seed to another value then it will fail the check */
        if (!TEST_int_ne(EC_GROUP_check_named_curve(group, 0), 0))
            goto err;
    } else {
        /* If there is no record of a seed for a built in curve and we set the seed to any value then it should pass the check (as the seed is optional and the library does not verify it) */
        if (!TEST_int_ne(EC_GROUP_check_named_curve(group, 0), nid))
            goto err;
    }
    /* In any case, pass if the seed is unknown (as it is optional) */
    if (!TEST_int_eq(EC_GROUP_set_seed(group, NULL, 0), 1)
            || !TEST_int_ne(EC_GROUP_check_named_curve(group, 0), nid))
        goto err;

    // [...]
~~~

Reply:
Changed... thanks